README.md
## ***`argone`***

---

Most intuitive global cli maker. *(lazy_static + config-rs + clap)


[![CI][ci-badge]][ci-url]
[![Crates.io][crates-badge]][crates-url]
[![Licensed][license-badge]][license-url]
[![Twitter][twitter-badge]][twitter-url]

[ci-badge]: https://github.com/just-do-halee/argone/actions/workflows/rust.yml/badge.svg
[crates-badge]: https://img.shields.io/crates/v/argone.svg?labelColor=383636
[license-badge]: https://img.shields.io/crates/l/argone?labelColor=383636
[twitter-badge]: https://img.shields.io/twitter/follow/do_halee?style=flat&logo=twitter&color=4a4646&labelColor=333131&label=just-do-halee

[ci-url]: https://github.com/just-do-halee/argone/actions
[twitter-url]: https://twitter.com/do_halee
[crates-url]: https://crates.io/crates/argone
[license-url]: https://github.com/just-do-halee/argone
| [Examples](https://github.com/just-do-halee/argone/tree/main/examples) | [Docs](https://docs.rs/argone) | [Latest Note](https://github.com/just-do-halee/argone/blob/main/CHANGELOG.md) | 

```toml
[dependencies]
argone = "0.5"
```

---

## *Phases*

1. Parsing `clap`.
2. If clap-argument is `None`<br>and the argument has `[Config]` mark<br>then extract data from `Config(file or env)`.
3. When the data is extracted and that was empty,<br>eventually set Default(=) value.

---

# `Example`
##### * ***basics***
```rust
use argone::{prelude::*, *};

ARGONE! {

    version = "0.1"
    author = "just-do-halee <just.do.halee@gmail.com>"
    about = "this is our application"

    Config {
        // file's extension could be
        // ini | json | json5 | yaml | toml | ron
        file = "examples/Config"
        // arg's env-prefix ex:) APP_...
        prefix = "APP"
    }

    Args {

        /// This is just one required argument.
        /// and three slashes will be an 'about' for clap cli.
        name: String

        /// This argument is connected to Config(file or env).
        /// Template is
        /// [Config] $name: Option<$type> = Some($default_value) or
        /// [Config] $name: Vec<$type> = vec![$default_values]
        /// and this works in non-required argument.
        [Config] age: Option<u8> = Some(12),
        [Config] job: Vec<String>

        // And all arugments are working on clap_derive format.

        /// Exactly same with clap_derive.
        (short, long)
        parents: Vec<String>

        /// But (default_value = "..") doesn't work to
        /// the config arguments. instead, this would work.
        (short, long, name = "WEIGHT")
        [Config] weight: Option<u8> = Some(50)

        /// In normal arguments, default_value will work.
        (short, long, default_value = "1")
        verbose: u8

    }

    commands = Sub

}

COMMANDS! {
    Sub {

        /// The subcommand
        /// but subcommands do not have config arguments.
        First {
            version = "1.0"
            author = "just-do-halee <just.do.halee@gmail.com>"
            Args {
                /// This area is also same.
                test: String
            }
        }

        Second {
            about = "The second subcommand"
            Args {
                test: u8
            }
        }

    }
}

fn main() {
    println!("{:#?}", *ARGS);

    if ARGS.name == "test" {
        println!("name is test.");
    }

    if ARGS.age.unwrap() != 12 {
        println!("age is not matching default 12.");
    }

    for job in &ARGS.job {
        println!("job: {}", job);
    }

    for parents in &ARGS.parents {
        println!("parent: {}", parents);
    }

    println!(
        "weight is {}\nverbose is {}",
        ARGS.weight.unwrap(),
        ARGS.verbose
    );

    if let Some(sub) = &ARGS.commands {
        match sub {
            Sub::First { test } => println!("first command: {}", test),
            Sub::Second { test } => println!("second command: {}", test),
        }
    } else {
        println!("none subcommands");
    }
}
```


CHANGELOG.md
## 0.5.2 (June 19, 2022)

### Release 0.5.2

- edition = "2021"

## 0.5.1 (June 19, 2022)

### Release 0.5.1

- clap = { version = "3.2.5", features = ["derive"] }
- fixing clap bug

---

## 0.5.0 (December 3, 2021)

### Release 0.5.0

- To prevent a conflict
- between the macro name and the structure name,
- the macro was renamed `ARGONE`.

---

## 0.4.0 (December 3, 2021)

### Release 0.4.0

- removing unnecessary #[clap()] tokens.
- now [Config] can parse Vector types.
- fixing small bugs.

[Example]

- (short, long)
- [Config] someList: Vec\<String\> = vec!["default".to_string()]

---

## 0.3.0 (December 2, 2021)

### Release 0.3.0

- adding prelude module.
- please use<br>
  `pub use argone::{*, prelude::*};`<br>or<br>`pub use argone::{..., prelude::*};`

---

## 0.2.0 (December 2, 2021)

### Release 0.2.0


src/lib.rs
// Licensed under either of Apache License, Version 2.0 or MIT license at your option.
// Copyright 2021 Hwakyeom Kim(=just-do-halee)

//! # **`argone`**
//!
//! Most intuitive global cli maker. *(lazy_static + config-rs + clap)<br>
//!
//! <a href="https://github.com/just-do-halee/argone/tree/main/examples">Examples</a>

#![allow(unused)]

#[macro_export]
macro_rules! ARGONE {
    (@PassOrB $a:expr, $b:expr) => {
        $a
    };
    (@PassOrB , $b:expr) => {
        $b
    };
    (@Parse) => {
        Args::parse()
    };
    (@Parse $file:literal, $($prefix:literal)?, $( ($($config_panic:expr),*) )?) => {
        Args::parse_with_config({
            let mut conf = config::Config::default();
            // ini | json | json5 | yaml | toml | ron
            // arg's prefix must be '$prefix'. ex:) LOX_...
            ARGONE!(@SetConfig conf, $file, $($prefix)?, $( ($($config_panic),*) )?)
        })
    };
    (@SetConfig $config:expr, $file:literal, ,) => {
        {
            $config.merge(config::File::with_name($file)).ok();
            $config
        }
    };
    (@SetConfig $config:expr, $file:literal, , ($($config_panic:expr),*)) => {
        {
            $config.merge(config::File::with_name($file))
            .unwrap_or_else(|_| panic!($($config_panic),*))
            $config
        }
    };
    (@SetConfig $config:expr, $file:literal, $prefix:literal,) => {
        {
            $config.merge(config::File::with_name($file)).ok();
            $config.merge(config::Environment::with_prefix($prefix)).ok();
            $config
        }
    };
    (@SetConfig $config:expr, $file:literal, $prefix:literal, ($($config_panic:expr),*)) => {
        {
            $config.merge(config::File::with_name($file))
            .unwrap_or_else(|_| panic!($($config_panic),*))
            .merge(config::Environment::with_prefix($prefix))
            .expect("init");
            $config
        }
    };

    (@SetConfig @If $conf:expr, $opts:expr, $name:ident,)
    => {
    };
    (@SetConfig @If $conf:expr, $opts:expr, [Config] $name:ident, $($default:expr)?)
    => {
        if $opts.$name.is_empty_or_none() {
             $opts.$name = if let Ok(v) = $conf.get(stringify!($name)) {
                                v
                            }
                            else {
                                ARGONE!(@PassOrB $($default)?, $opts.$name)
                            };
        }
    };


    (
        @Args {
            {
                $(#[$meta:meta])*
            }
            $(
                $(#[$f_meta:meta])*
                $([$tt:tt])? $name:ident: $ty:ty$( = $default:expr)?,
            )*
        }
    ) => {
        #[allow(non_snake_case)]
        #[derive(Debug, Parser)]
        $(#[$meta])*
        pub struct Args {
            $(
                $(#[$f_meta])*
                pub $name: $ty,
            )*
        }
        impl Args {
            pub fn parse_with_config(conf: config::Config) -> Self {
                let mut opts = Args::parse();
                $(
                    ARGONE!(@SetConfig @If conf, opts, $([$tt])? $name, $($default)?);
                )*
                opts
            }
        }
    };

    (
        $(#[$main_meta:meta])*
        $(version = $version:literal $(;)?$(,)?)?
        $(author = $author:literal $(;)?$(,)?)?
        $(about = $about:literal $(;)?$(,)?)?
        $(Config {
            file = $config_file:literal $(;)?$(,)?
            $(prefix = $config_prefix:literal $(;)?$(,)?)?
            $(panic = ($($config_panic:expr),*) $(;)?$(,)?)?
        })?
        Args {
            $(
                $(#[$args_meta:meta])*
                $(($($clap_arg:stmt),*))?
                $([$tt:tt])? $args_name:ident: $args_ty:ty $(= $args_default:expr)? $(;)?$(,)?
            )+
        }
        $(
            $(#[$subcommands_meta:meta])*
            commands = $subcommands:ty $(;)?$(,)?
        )?
    ) => {

        ARGONE!(
            @Args {
                {
                    $(#[$main_meta])*
                    #[clap($(version = $version,)? $(author = $author,)? $(about = $about,)?)]
                }
                $(
                    $(#[$args_meta])*
                    $(#[clap($($clap_arg),*)])?
                    $([$tt])? $args_name: $args_ty$(= $args_default)?,
                )+
                $(
                    $(#[$subcommands_meta])*
                    #[clap(subcommand)]
                    commands: Option<$subcommands>,
                )?
            }
        );

        lazy_static! {
            pub static ref ARGS: Args = ARGONE!(@Parse $($config_file, $($config_prefix)?, $( ($($config_panic),*) )?)?);
        }

    };
}

#[macro_export]
macro_rules! COMMANDS {
    (
        $(
            $(#[$main_meta:meta])*
            $name:ident {
            $(
                $(#[$meta:meta])*
                $(($($clap_arg:stmt),*))?
                $command_name:ident {
                $(version = $version:literal)?
                $(author = $author:literal)?
                $(about = $about:literal)?
                $(
                    Args {
                        $(
                            $(#[$args_meta:meta])*
                            $args_name:ident: $args_ty:ty
                        )*
                    }
                )?

                $(
                    $(#[$subcommands_meta:meta])*
                    commands = $subcommands:ty
                )?
            })+
        })+
    ) => {

        $(#[allow(non_snake_case)]
        #[derive(Debug, Clone, PartialEq, Eq, Parser)]
        $(#[$main_meta])*
        pub enum $name {
            $(
                $(#[$meta])*
                $(#[clap($($clap_arg),*)])?
                #[clap($(version = $version,)? $(author = $author,)? $(about = $about)?)]
                $command_name {
                    $($(
                        $(#[$args_meta])*
                        $args_name: $args_ty,
                    )*)?
                    $(
                        $(#[$subcommands_meta])*
                        #[clap(subcommand)]
                        commands: Option<$subcommands>,
                    )?
                },
            )+
        })+

    };
}

pub mod prelude {
    pub use clap::{self, Parser};
    pub use config;
    pub use lazy_static::lazy_static;

    use std::any::Any;
    pub trait ArgOne {
        fn is_empty_or_none(&self) -> bool;
    }

    impl<T> ArgOne for Option<T> {
        #[inline]
        fn is_empty_or_none(&self) -> bool {
            self.is_none()
        }
    }

    impl<T> ArgOne for Vec<T> {
        #[inline]
        fn is_empty_or_none(&self) -> bool {
            self.is_empty()
        }
    }
}

use std::{env, path::PathBuf};

prelude::lazy_static! {
    pub static ref CURRENT_EXE: PathBuf = env::current_exe()
        .unwrap()
        .file_name()
        .map(PathBuf::from)
        .unwrap();
    pub static ref CURRENT_DIR: PathBuf = env::current_dir().unwrap();
}


examples/advanced.rs
use argone::{prelude::*, *};
use std::path::PathBuf;

ARGONE! {

    /// this area is top of the clap struct meta area.
    #[derive(Clone)]
    #[clap(after_help = "My Application")]

    version = "0.1"
    author = "just-do-halee <just.do.halee@gmail.com>"

    Config {
        file = "examples/Config"
        prefix = "APP"
        panic = ("couldn't find {} file.", "Config")
    }

    Args {
        /// Root directory
        [Config] rootDir: Option<PathBuf> = Some(CURRENT_DIR.clone())

        /// This name will be changed as app-name or APP_NAME
        (short, long, default_value = "test")
        appName: String; // can be separated by ';' or ','

        /// Some List
        (short, long)
        [Config] someList: Vec<String> = vec!["default".to_string()]

        /// A level of verbosity, and can be used multiple times
        (short, long, parse(from_occurrences))
        verbose: u8,
    }

    commands = Sub2
}

COMMANDS! {
    /// this area is top of the command enum meta area.
    #[derive(Copy)]
    Sub {
        /// third
        Third {
            Args {
                // Test u8
                test: u8
            }
        }
    }

    /// same
    Sub2 {
        /// first
        (alias = "foo")
        First {
            version = "1.0"
            author = "just-do-halee <just.do.halee@gmail.com>"
            Args {
                /// Test String
                test: String
            }
        }
        /// second
        (aliases = &["baz", "fizz"])
        Second {
            Args {
                // Test u8
                test: u8
            }
        }
    }
}

fn main() {
    println!("{:#?}", *ARGS);
    println!("{}", CURRENT_EXE.display());
}


examples/basic.rs
use argone::{prelude::*, *};

ARGONE! {

    version = "0.1"
    author = "just-do-halee <just.do.halee@gmail.com>"
    about = "this is our application"

    Config {
        // file's extension could be
        // ini | json | json5 | yaml | toml | ron
        file = "examples/Config"
        // arg's env-prefix ex:) APP_...
        prefix = "APP"
    }

    Args {

        /// This is just one required argument.
        /// and three slashes will be an 'about' for clap cli.
        name: String

        /// This argument is connected to Config(file or env).
        /// Template is
        /// [Config] $name: Option<$type> = Some($default_value) or
        /// [Config] $name: Vec<$type> = vec![$default_values]
        /// and this works in non-required argument.
        [Config] age: Option<u8> = Some(12),
        [Config] job: Vec<String>

        // And all arugments are working on clap_derive format.

        /// Exactly same with clap_derive.
        (short, long)
        parents: Vec<String>

        /// But (default_value = "..") doesn't work to
        /// the config arguments. instead, this would work.
        (short, long, name = "WEIGHT")
        [Config] weight: Option<u8> = Some(50)

        /// In normal arguments, default_value will work.
        (short, long, default_value = "1")
        verbose: u8

    }

    commands = Sub

}

COMMANDS! {
    Sub {

        /// The subcommand
        /// but subcommands do not have config arguments.
        First {
            version = "1.0"
            author = "just-do-halee <just.do.halee@gmail.com>"
            Args {
                /// This area is also same.
                test: String
            }
        }

        Second {
            about = "The second subcommand"
            Args {
                test: u8
            }
        }

    }
}

fn main() {
    println!("{:#?}", *ARGS);

    if ARGS.name == "test" {
        println!("name is test.");
    }

    if ARGS.age.unwrap() != 12 {
        println!("age is not matching default 12.");
    }

    for job in &ARGS.job {
        println!("job: {}", job);
    }

    for parents in &ARGS.parents {
        println!("parent: {}", parents);
    }

    println!(
        "weight is {}\nverbose is {}",
        ARGS.weight.unwrap(),
        ARGS.verbose
    );

    if let Some(sub) = &ARGS.commands {
        match sub {
            Sub::First { test } => println!("first command: {}", test),
            Sub::Second { test } => println!("second command: {}", test),
        }
    } else {
        println!("none subcommands");
    }
}
